sam/assembly.go

```diff
@@ -0,0 +1,219 @@
+// a diploid base at that position (in other words, not considering INDELs).
+// epsilon represents the error rate, a parameter for the likelihood function.
+func DiploidBaseCallFromPile(p Pile, refBase dna.Base, priorCache [][]float64, homozygousCache [][]float64, heterozygousCache [][]float64, epsilon float64) DiploidBase {
+	if refBase == dna.N { //we will force a return of diploid NN whenever an N is found in the reference.
+		return NN
+	}
+	var aCount = p.CountF[dna.A] + p.CountR[dna.A]
+	var cCount = p.CountF[dna.C] + p.CountR[dna.C]
+	var gCount = p.CountF[dna.G] + p.CountR[dna.G]
+	var tCount = p.CountF[dna.T] + p.CountR[dna.T]
+	var baseCount = aCount + cCount + gCount + tCount
+
+	if baseCount < 1 {
+		switch refBase {
+		case dna.A:
+			return AA
+		case dna.C:
+			return CC
+		case dna.G:
+			return GG
+		case dna.T:
+			return TT
+		case dna.N: //should be redundant with catch above
+			return NN
+		default:
+			log.Fatalf("Reference base was not N, A, C, G, or T. Found: %s.\n", dna.BaseToString(refBase))
+		}
+	}
+
+	var maxDiploid []DiploidBase
+	var maxPosterior float64
+
+	switch refBase {
+	case dna.A:
+		maxDiploid = []DiploidBase{AA}
+		maxPosterior = logspace.Multiply(baseLikelihood(aCount, cCount, gCount, tCount, AA, epsilon, homozygousCache, heterozygousCache), priorCache[dna.A][AA])
+	case dna.C:
+		maxDiploid = []DiploidBase{CC}
+		maxPosterior = logspace.Multiply(baseLikelihood(aCount, cCount, gCount, tCount, CC, epsilon, homozygousCache, heterozygousCache), priorCache[dna.C][CC])
+	case dna.G:
+		maxDiploid = []DiploidBase{GG}
+		maxPosterior = logspace.Multiply(baseLikelihood(aCount, cCount, gCount, tCount, GG, epsilon, homozygousCache, heterozygousCache), priorCache[dna.G][GG])
+	case dna.T:
+		maxDiploid = []DiploidBase{TT}
+		maxPosterior = logspace.Multiply(baseLikelihood(aCount, cCount, gCount, tCount, TT, epsilon, homozygousCache, heterozygousCache), priorCache[dna.T][TT])
+	default:
+		log.Fatalf("Reference base was not N, A, C, G, or T. Found: %s.\n", dna.BaseToString(refBase))
+	}
+
+	//DEBUG: fmt.Printf("HomoRefBase: %s. Posterior: %v.\n", diploidBaseString(maxDiploid[0]), maxPosterior)
+
+	var geno DiploidBase
+	var currPosterior float64
+	for geno = 0; geno < 10; geno++ { //for each genotype, encoded as a DiploidBase byte ranging from 0 to 9, inclusive
+		currPosterior = logspace.Multiply(baseLikelihood(aCount, cCount, gCount, tCount, geno, epsilon, homozygousCache, heterozygousCache), priorCache[refBase][geno])
+		fmt.Printf("CurrGeno: %s. CurrPosterior: %v.\n", diploidBaseString(geno), currPosterior)
+		if currPosterior > maxPosterior {
+			maxPosterior = currPosterior
+			maxDiploid = maxDiploid[:1] //clear ties
+			maxDiploid[0] = geno
+		} else if currPosterior == maxPosterior {
+			maxDiploid = append(maxDiploid, geno)
+		}
+	}
+
+	return maxDiploid[numbers.RandIntInRange(0, len(maxDiploid))] //if two genotypes have the same posterior density, pick one at random.
```

Comment:
(For future cmd) Should there be a note in usage about "if two genotypes have the same posterior density, pick one at random"?